plugins/analytics-module-ga/src/apis/implementations/AnalyticsApi/GoogleAnalytics.ts

```diff
@@ -151,4 +193,61 @@ export class GoogleAnalytics implements AnalyticsApi {
 
     return customDimensionsMetrics;
   }
+
+  /**
+   * Sets the GA userId, based on the `userEntityRef` set on the backstage
+   * identity loaded from a given Backstage Identity API instance. Because
+   * Google forbids sending any PII (including on the userId field), we hash
+   * the entire `userEntityRef` on behalf of integrators:
+   *
+   * - With value `User:default/name`, userId becomes `sha256(User:default/name)`
+   *
+   * If an integrator wishes to use an alternative hashing mechanism or an
+   * entirely different value, they may do so by passing a dummy Identity API
+   * implementation which returns a `userEntityRef` whose kind is the literal
+   * string `PrivateUser`, whose namespace is anything (it will be ignored) and
+   * whose name is the pre-hashed ID value.
+   *
+   * - With value `PrivateUser:default/a0n3b4n3`, userId becomes `a0n3b4n3`
+   * - With `PrivateUser:xyz/a0n3b4n3`, userId is `a0n3b4n3`
+   *
+   * Note: this feature requires that an integrator has set up a Google
+   * Analytics User ID view in the property used to track Backstage.
+   */
+  private async setUserFrom(identityApi: IdentityApi) {
+    const { userEntityRef } = await identityApi.getBackstageIdentity();
+
+    // Prevent PII from being passed to Google Analytics.
+    const userId = await this.getPrivateUserId(userEntityRef);
+
+    // Set the user ID.
+    ReactGA.set({ userId });
+
+    // Notify the deferred capture mechanism that it may proceed.
+    this.capture.setReady();
+  }
+
+  /**
+   * Returns a PII-free user ID for use in Google Analytics.
+   */
+  private getPrivateUserId(userEntityRef: string): Promise<string> {
+    // Allow integrators to provide their own hashing transformer.
+    if (this.customUserIdTransform) {
+      return this.customUserIdTransform(userEntityRef);
+    }
+
+    return this.hash(userEntityRef);
+  }
+
+  /**
+   * Simple hash function; relies on web cryptography + the sha-256 algorithm.
+   */
+  private async hash(value: string): Promise<string> {
+    const digest = await crypto.subtle.digest(
+      'sha-256',
+      new TextEncoder().encode(value),
```

Comment:
Bit of a side-note here that this will be a breaking change for anyone that currently relies on just the user ID being hashed, as we're now hashing the entire ref. But perhaps that's why the custom transform is here to begin with :grin: